File: src/engine/plan_status.py
"""Plan status line management — update markdown status from engine state.

The **Status:** line in build plan markdown is derived output, updated by the
engine when convergence state changes. It is NEVER the source of truth.
"""

import os
import re


VALID_STATUSES = {"NOT STARTED", "IN PROGRESS", "CONVERGED", "ESCALATED"}


def update_plan_status(plan_file: str, new_status: str) -> bool:
    """Update the **Status:** line in a build plan markdown file.

    Returns True if the status was updated, False if no status line found
    or file doesn't exist. Never fails — if the line isn't there, we skip.
    """
    if new_status not in VALID_STATUSES:
        return False

    if not os.path.exists(plan_file):
        return False

    try:
        with open(plan_file) as f:
            content = f.read()
    except OSError:
        return False

    # Match **Status:** followed by any value
    pattern = r'(\*\*Status:\*\*\s*).+'
    match = re.search(pattern, content)
    if not match:
        return False

    updated = re.sub(pattern, rf'\g<1>{new_status}', content, count=1)
    if updated == content:
        return False  # Already at this status

    try:
        with open(plan_file, "w") as f:
            f.write(updated)
        return True
    except OSError:
        return False


def read_plan_status(plan_file: str) -> str | None:
    """Read the current **Status:** line from a plan file."""
    if not os.path.exists(plan_file):
        return None

    try:
        with open(plan_file) as f:
            content = f.read()
    except OSError:
        return None

    match = re.search(r'\*\*Status:\*\*\s*(.+)', content)
    if match:
        return match.group(1).strip()
    return None

File: rust/src/engine/plan_status.rs
//! Plan status line management — update markdown status from engine state.

use regex::Regex;
use std::fs;

const VALID_STATUSES: &[&str] = &["NOT STARTED", "IN PROGRESS", "CONVERGED", "ESCALATED"];

/// Update the **Status:** line in a plan file. Returns true if updated.
pub fn update_plan_status(plan_file: &str, new_status: &str) -> bool {
    if !VALID_STATUSES.contains(&new_status) {
        return false;
    }
    let content = match fs::read_to_string(plan_file) {
        Ok(c) => c,
        Err(_) => return false,
    };

    let re = Regex::new(r"(\*\*Status:\*\*\s*).+").unwrap();
    if !re.is_match(&content) {
        return false;
    }

    let updated = re.replace(&content, format!("${{1}}{new_status}")).to_string();
    if updated == content {
        return false;
    }

    fs::write(plan_file, &updated).is_ok()
}

/// Read the current **Status:** line from a plan file.
pub fn read_plan_status(plan_file: &str) -> Option<String> {
    let content = fs::read_to_string(plan_file).ok()?;
    let re = Regex::new(r"\*\*Status:\*\*\s*(.+)").unwrap();
    re.captures(&content).map(|c| c[1].trim().to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn update_status() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("plan.md");
        fs::write(&path, "# Plan\n**Status:** NOT STARTED\n").unwrap();
        assert!(update_plan_status(path.to_str().unwrap(), "IN PROGRESS"));
        assert!(fs::read_to_string(&path).unwrap().contains("IN PROGRESS"));
    }

    #[test]
    fn update_to_converged() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("plan.md");
        fs::write(&path, "# Plan\n**Status:** IN PROGRESS\n").unwrap();
        assert!(update_plan_status(path.to_str().unwrap(), "CONVERGED"));
    }

    #[test]
    fn no_status_line() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("plan.md");
        fs::write(&path, "# Plan\nNo status\n").unwrap();
        assert!(!update_plan_status(path.to_str().unwrap(), "IN PROGRESS"));
    }

    #[test]
    fn invalid_status() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("plan.md");
        fs::write(&path, "# Plan\n**Status:** NOT STARTED\n").unwrap();
        assert!(!update_plan_status(path.to_str().unwrap(), "INVALID"));
    }

    #[test]
    fn already_at_status() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("plan.md");
        fs::write(&path, "# Plan\n**Status:** IN PROGRESS\n").unwrap();
        assert!(!update_plan_status(path.to_str().unwrap(), "IN PROGRESS"));
    }

    #[test]
    fn read_status() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("plan.md");
        fs::write(&path, "# Plan\n**Status:** CONVERGED\n").unwrap();
        assert_eq!(read_plan_status(path.to_str().unwrap()), Some("CONVERGED".into()));
    }

    #[test]
    fn read_status_missing() {
        assert_eq!(read_plan_status("/nonexistent.md"), None);
    }

    #[test]
    fn file_not_found() {
        assert!(!update_plan_status("/nonexistent.md", "IN PROGRESS"));
    }
}
